File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& ppp) {
        vector<vector<int>> g(n,vector<int>());
        for(auto x:ppp) g[x[1]].push_back(x[0]);
        vector<int> d(n,0);
        function<bool(int)>dfs = [&](int u){
            d[u]=1;
            for(auto v:g[u]){
                if(d[v] == 0){
                    if(dfs(v))
                        return true;
                }
                else if(d[v]==1){
                    return true;
                }
            }
            d[u]=2;
            return false;
        };
        for(int i = 0; i <n; i++){
            if(d[i]==0)
                if(dfs(i)==true)
                    return !true;
        }
        return !false;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, n: int, pr: List[List[int]]) -> bool:
        g = [[] for _ in range(n)]
        ind = [0]*n
        for p in pr:
            g[p[0]].append(p[1])
            ind[p[1]]+=1
        q = deque()
        for i in range(n):
            if ind[i]==0:
                q.append(i)
        while q:
            u = q.popleft()
            for v in g[u]:
                ind[v]-=1
                if ind[v]==0:
                    q.append(v)
        for i in range(n):
            if ind[i]>0:
                return False
        return True
        